File: app/src/main/java/com/example/tecreciclaje/userpanel/PerfilActivity.java
package com.example.tecreciclaje.userpanel;

import android.app.Dialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.canhub.cropper.CropImageContract;
import com.canhub.cropper.CropImageContractOptions;
import com.canhub.cropper.CropImageOptions;
import com.canhub.cropper.CropImageView;
import com.example.tecreciclaje.LoginActivity;
import com.example.tecreciclaje.Model.CircleTransform;
import com.example.tecreciclaje.R;
import com.example.tecreciclaje.UserPanel;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.squareup.picasso.Picasso;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class PerfilActivity extends AppCompatActivity {

    private FirebaseAuth auth;
    private DatabaseReference userRef;
    private FirebaseStorage storage;
    private StorageReference storageRef;

    private Button btnActualizar, btnSelectImage;
    private EditText editTextNombre, editTextApellido, editTextCorreo, editTextEdad;
    private ImageView imageViewProfile;

    private Uri imageUri;
    private String imageUrlActual;
    private Dialog loadingDialog;

    private final ActivityResultLauncher<CropImageContractOptions> cropImage =
            registerForActivityResult(new CropImageContract(), result -> {
                if (result.isSuccessful()) {
                    Uri uriContent = result.getUriContent();
                    imageUri = uriContent;
                    imageViewProfile.setImageURI(imageUri);
                } else {
                    Toast.makeText(this, "Error al recortar imagen.", Toast.LENGTH_SHORT).show();
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_perfil);

        auth = FirebaseAuth.getInstance();
        storage = FirebaseStorage.getInstance();
        storageRef = storage.getReference();

        initializeViews();
        setupNavigation();
        setupLoadingDialog();

        FirebaseUser currentUser = auth.getCurrentUser();
        if (currentUser == null) {
            startActivity(new Intent(PerfilActivity.this, LoginActivity.class));
            finish();
            return;
        }

        userRef = FirebaseDatabase.getInstance().getReference("usuarios").child(currentUser.getUid());
        getUserData();

        btnActualizar.setOnClickListener(v -> confirmUpdateProfile());
        btnSelectImage.setOnClickListener(v -> showImageSourceDialog());
    }

    private void initializeViews() {
        btnActualizar = findViewById(R.id.btnActualizar);
        btnSelectImage = findViewById(R.id.btnSelectImage);

        editTextNombre = findViewById(R.id.editTextNombre);
        editTextApellido = findViewById(R.id.editTextApellido);
        editTextCorreo = findViewById(R.id.editTextCorreo);
        editTextEdad = findViewById(R.id.editTextEdad);
        imageViewProfile = findViewById(R.id.imageViewProfile2);
    }

    private void setupNavigation() {
        BottomNavigationView bottomNavigationView = findViewById(R.id.bottom_navigation);
        bottomNavigationView.setSelectedItemId(R.id.nav_perfil);

        bottomNavigationView.setOnItemSelectedListener(item -> {
            int itemId = item.getItemId();
            Class<?> currentClass = getClass();

            if (itemId == R.id.nav_home && !currentClass.equals(UserPanel.class)) {
                startActivity(new Intent(this, UserPanel.class));
                overridePendingTransition(0, 0);
                return true;

            } else if (itemId == R.id.nav_docs && !currentClass.equals(MisValesActivity.class)) {
                startActivity(new Intent(this, MisValesActivity.class));
                overridePendingTransition(0, 0);
                return true;

            } else if (itemId == R.id.nav_histori && !currentClass.equals(HistorialActivity.class)) {
                startActivity(new Intent(this, HistorialActivity.class));
                overridePendingTransition(0, 0);
                return true;

            } else if (itemId == R.id.nav_perfil && !currentClass.equals(PerfilActivity.class)) {
                startActivity(new Intent(this, PerfilActivity.class));
                overridePendingTransition(0, 0);
                return true;
            }

            return true;
        });
    }

    private void setupLoadingDialog() {
        loadingDialog = new Dialog(this);
        loadingDialog.setContentView(R.layout.dialog_loading);
        loadingDialog.setCancelable(false);
        loadingDialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);
        loadingDialog.getWindow().setLayout(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
    }

    private void showImageSourceDialog() {
        String[] options = {"Cámara", "Galería"};
        new AlertDialog.Builder(this)
                .setTitle("Selecciona la fuente de la imagen")
                .setItems(options, (dialog, which) -> {
                    if (which == 0) {
                        Intent cameraIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
                        startActivityForResult(cameraIntent, 100);
                    } else {
                        selectImage();
                    }
                })
                .show();
    }

    private void selectImage() {
        Intent intent = new Intent(Intent.ACTION_PICK);
        intent.setType("image/*");
        startActivityForResult(Intent.createChooser(intent, "Selecciona una imagen"), 200);
    }

    private void getUserData() {
        userRef.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot snapshot) {
                if (snapshot.exists()) {
                    String nombre = snapshot.child("nombre").getValue(String.class);
                    String apellido = snapshot.child("apellido").getValue(String.class);
                    String correo = snapshot.child("email").getValue(String.class);
                    String edad = snapshot.child("edad").getValue(String.class);
                    imageUrlActual = snapshot.child("perfil").getValue(String.class);

                    editTextNombre.setText(nombre);
                    editTextApellido.setText(apellido);
                    editTextCorreo.setText(correo);
                    editTextEdad.setText(edad);

                    if (imageUrlActual != null) {
                        Picasso.get().load(imageUrlActual).transform(new CircleTransform()).into(imageViewProfile);
                    }
                }
            }

            @Override
            public void onCancelled(DatabaseError error) {
                Toast.makeText(PerfilActivity.this, "Error al cargar datos", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void confirmUpdateProfile() {
        new AlertDialog.Builder(this)
                .setTitle("Actualizar Perfil")
                .setMessage("¿Deseas guardar los cambios?")
                .setPositiveButton("Sí", (dialog, which) -> updateUserProfile())
                .setNegativeButton("No", (dialog, which) -> dialog.dismiss())
                .show();
    }

    private void updateUserProfile() {
        loadingDialog.show();
        String nombre = editTextNombre.getText().toString().trim();
        String apellido = editTextApellido.getText().toString().trim();
        String correo = editTextCorreo.getText().toString().trim();
        String edad = editTextEdad.getText().toString().trim();

        if (nombre.isEmpty() || apellido.isEmpty() || correo.isEmpty() || edad.isEmpty()) {
            loadingDialog.dismiss();
            Toast.makeText(this, "Completa todos los campos.", Toast.LENGTH_SHORT).show();
            return;
        }

        if (imageUri != null) {
            uploadImageAndSave(nombre, apellido, correo, edad);
        } else {
            guardarDatosEnDatabase(nombre, apellido, correo, edad, imageUrlActual);
        }
    }

    private void uploadImageAndSave(String nombre, String apellido, String correo, String edad) {
        StorageReference imgRef = storageRef.child("Perfiles/" + UUID.randomUUID().toString());
        imgRef.putFile(imageUri).addOnSuccessListener(taskSnapshot ->
                imgRef.getDownloadUrl().addOnSuccessListener(uri -> {
                    guardarDatosEnDatabase(nombre, apellido, correo, edad, uri.toString());
                })
        ).addOnFailureListener(e -> {
            loadingDialog.dismiss();
            Toast.makeText(this, "Error al subir imagen.", Toast.LENGTH_SHORT).show();
        });
    }

    private void guardarDatosEnDatabase(String nombre, String apellido, String correo, String edad, String imagenUrl) {
        Map<String, Object> datos = new HashMap<>();
        datos.put("nombre", nombre);
        datos.put("apellido", apellido);
        datos.put("email", correo);
        datos.put("edad", edad);
        if (imagenUrl != null) datos.put("perfil", imagenUrl);

        userRef.updateChildren(datos).addOnSuccessListener(aVoid -> {
            loadingDialog.dismiss();
            Toast.makeText(this, "Perfil actualizado.", Toast.LENGTH_SHORT).show();
            startActivity(new Intent(this, UserPanel.class));
            finish();
        }).addOnFailureListener(e -> {
            loadingDialog.dismiss();
            Toast.makeText(this, "Error al actualizar perfil.", Toast.LENGTH_SHORT).show();
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_OK) {
            if (requestCode == 100) {
                Uri cameraImageUri = data.getData();
                if (cameraImageUri != null) {
                    launchCropper(cameraImageUri);
                } else if (data.getExtras() != null) {
                    imageUri = (Uri) data.getExtras().get("data");
                    launchCropper(imageUri);
                }
            } else if (requestCode == 200 && data != null) {
                Uri selectedImage = data.getData();
                if (selectedImage != null) {
                    launchCropper(selectedImage);
                }
            }
        }
    }

    private void launchCropper(Uri imageUri) {
        CropImageOptions options = new CropImageOptions();
        options.cropShape = CropImageView.CropShape.OVAL;
        options.aspectRatioX = 1;
        options.aspectRatioY = 1;
        options.fixAspectRatio = true;

        cropImage.launch(new CropImageContractOptions(imageUri, options));
    }
}

File: app/src/main/java/com/example/tecreciclaje/userpanel/PerfilActivity.kt
package com.example.tecreciclaje.userpanel

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.tecreciclaje.LoginActivity
import com.example.tecreciclaje.R
import com.example.tecreciclaje.utils.CircleTransform
import com.example.tecreciclaje.utils.FCMTokenManager
import com.example.tecreciclaje.utils.TutorialManager
import com.example.tecreciclaje.Model.UpdateNfcBottomSheetDialogFragment
import com.example.tecreciclaje.UserPanelDynamic
import com.example.tecreciclaje.utils.AppLogger
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.squareup.picasso.Picasso

class PerfilActivity : AppCompatActivity() {

    private lateinit var auth: FirebaseAuth
    private lateinit var userRef: DatabaseReference

    private lateinit var textViewUserName: TextView
    private lateinit var textViewUserEmail: TextView
    private lateinit var textViewUserCarrera: TextView
    private lateinit var textViewUserNumControl: TextView
    private lateinit var textViewUserEdad: TextView
    private lateinit var imageViewProfile2: ImageView
    private lateinit var optionEditProfile: View
    private lateinit var optionUpdateNfc: View
    private lateinit var optionVerNip: View
    private lateinit var textViewNip: TextView
    private lateinit var optionVerLogros: View

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_perfil)

        auth = FirebaseAuth.getInstance()
        initializeViews()
        setupNavigation()
        setupClickListeners()

        val currentUser = auth.currentUser
        if (currentUser == null) {
            startActivity(Intent(this, LoginActivity::class.java))
            finish()
            return
        }
        
        // MOSTRAR TUTORIAL SI ES NECESARIO
        TutorialManager.showPerfilTutorialIfNeeded(this)

        // VALIDAR TOKEN FCM AL ABRIR PERFIL
        FCMTokenManager.validateCurrentToken()

        userRef = FirebaseDatabase.getInstance().getReference("usuarios").child(currentUser.uid)
        loadUserData()
    }

    private fun initializeViews() {
        textViewUserName = findViewById(R.id.textViewUserName)
        textViewUserEmail = findViewById(R.id.textViewUserEmail)
        textViewUserCarrera = findViewById(R.id.textViewUserCarrera)
        textViewUserNumControl = findViewById(R.id.textViewUserNumControl)
        textViewUserEdad = findViewById(R.id.textViewUserEdad)
        imageViewProfile2 = findViewById(R.id.imageViewProfile2)
        
        optionEditProfile = findViewById(R.id.optionEditProfile)
        optionUpdateNfc = findViewById(R.id.optionUpdateNfc)
        optionVerNip = findViewById(R.id.optionVerNip)
        textViewNip = findViewById(R.id.textViewNip)
        optionVerLogros = findViewById(R.id.optionVerLogros)
    }

    private fun setupNavigation() {
        val bottomNavigationView = findViewById<BottomNavigationView>(R.id.bottom_navigation)
        bottomNavigationView.selectedItemId = R.id.nav_perfil

        bottomNavigationView.setOnItemSelectedListener { item ->
            val itemId = item.itemId
            val currentClass = this::class.java

            when {
                itemId == R.id.nav_home && currentClass != UserPanelDynamic::class.java -> {
                    startActivity(Intent(this, UserPanelDynamic::class.java))
                    overridePendingTransition(0, 0)
                    true
                }
                itemId == R.id.nav_docs && currentClass != MisValesActivity::class.java -> {
                    startActivity(Intent(this, MisValesActivity::class.java))
                    overridePendingTransition(0, 0)
                    true
                }
                itemId == R.id.nav_histori && currentClass != HistorialActivity::class.java -> {
                    startActivity(Intent(this, HistorialActivity::class.java))
                    overridePendingTransition(0, 0)
                    true
                }
                itemId == R.id.nav_perfil && currentClass != PerfilActivity::class.java -> {
                    startActivity(Intent(this, PerfilActivity::class.java))
                    overridePendingTransition(0, 0)
                    true
                }
                else -> false
            }
        }
    }

    private fun setupClickListeners() {
        // Opción: Editar Perfil
        optionEditProfile.setOnClickListener {
            val intent = Intent(this, EditarPerfilActivity::class.java)
            startActivity(intent)
        }

        // Opción: Actualizar NFC
        optionUpdateNfc.setOnClickListener {
            val dialog = UpdateNfcBottomSheetDialogFragment()
            dialog.show(supportFragmentManager, "UpdateNfcDialog")
        }

        // Opción: Ver NIP (copiar al portapapeles)
        optionVerNip.setOnClickListener {
            val nip = textViewNip.text.toString().replace("NIP: ", "")
            if (nip.isNotEmpty() && nip != "--------") {
                val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                val clip = ClipData.newPlainText("NIP", nip)
                clipboard.setPrimaryClip(clip)
                Toast.makeText(this, "NIP copiado al portapapeles", Toast.LENGTH_SHORT).show()
            }
        }

        // Opción: Ver Logros
        optionVerLogros.setOnClickListener {
            val intent = Intent(this, LogrosActivity::class.java)
            startActivity(intent)
        }
    }

    private fun loadUserData() {
        userRef.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.exists()) {
                    // Cargar información del usuario
                    val nombre = snapshot.child("usuario_nombre").getValue(String::class.java)
                    val apellido = snapshot.child("usuario_apellido").getValue(String::class.java)
                    val email = snapshot.child("usuario_email").getValue(String::class.java)
                    val imageUrl = snapshot.child("usuario_perfil").getValue(String::class.java)
                    val carrera = snapshot.child("usuario_carrera").getValue(String::class.java)
                    val numControl = snapshot.child("usuario_numControl").getValue(String::class.java)
                    val edad = snapshot.child("usuario_edad").getValue(String::class.java)
                    val nfcUid = snapshot.child("usuario_nfcUid").getValue(String::class.java) ?: ""
                    val nip = snapshot.child("usuario_nip").getValue(String::class.java) ?: ""

                    // Actualizar UI
                    val nombreCompleto = "${nombre ?: ""} ${apellido ?: ""}".trim()
                    textViewUserName.text = if (nombreCompleto.isEmpty()) "Usuario" else nombreCompleto
                    textViewUserEmail.text = email ?: "usuario@email.com"
                    
                    // Actualizar nuevos campos
                    textViewUserCarrera.text = "Carrera: ${carrera ?: "No especificada"}"
                    textViewUserNumControl.text = "Número de Control: ${numControl ?: "No especificado"}"
                    textViewUserEdad.text = "Edad: ${if (edad != null) "$edad años" else "No especificada"}"

                    // Mostrar/ocultar módulos según si tiene NFC o NIP
                    if (nfcUid.isEmpty() && nip.isNotEmpty()) {
                        // Usuario sin NFC pero con NIP - mostrar módulo NIP
                        optionVerNip.visibility = View.VISIBLE
                        optionUpdateNfc.visibility = View.GONE
                        textViewNip.text = "NIP: $nip"
                    } else if (nfcUid.isNotEmpty()) {
                        // Usuario con NFC - mostrar módulo NFC, ocultar NIP
                        optionVerNip.visibility = View.GONE
                        optionUpdateNfc.visibility = View.VISIBLE
                    } else {
                        // Usuario sin NFC ni NIP - ocultar ambos
                        optionVerNip.visibility = View.GONE
                        optionUpdateNfc.visibility = View.GONE
                    }

                    // Cargar imagen de perfil
                    if (!imageUrl.isNullOrEmpty()) {
                        Picasso.get()
                            .load(imageUrl)
                            .transform(CircleTransform())
                            .placeholder(R.drawable.user)
                            .error(R.drawable.user)
                            .into(imageViewProfile2)
                    } else {
                        imageViewProfile2.setImageResource(R.drawable.user)
                    }
                }
            }

            override fun onCancelled(error: DatabaseError) {
                AppLogger.e("Error al cargar datos del perfil: ${error.message}")
            }
        })
    }
}
